samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessor.java

```diff
@@ -0,0 +1,291 @@
+      TestZkStreamProcessorBase.TestStreamTask.endLatch.await(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      Assert.fail("endLatch.await failed with an interruption:" + e.getLocalizedMessage());
+    }
+
+    // collect all the threads
+    try {
+      for (Thread t : threads) {
+        synchronized (t) {
+          t.notify(); // to stop the thread
+        }
+        t.join(1000);
+      }
+    } catch (InterruptedException e) {
+      Assert.fail("Failed to join finished thread:" + e.getLocalizedMessage());
+    }
+
+    verifyNumMessages(outputTopic, messageCount, messageCount);
+  }
+
+  @Test
+  /**
+   * Similar to the previous tests, but add another processor in the middle
+   */ public void testStreamProcessorWithAdd() {
+
+    // set number of events we expect wo read by both processes in total:
+    // p1 - reads 'messageCount' at first
+    // p1 and p2 read all messageCount together, since they start from the beginning.
+    // so we expect total 3 x messageCounts
+    int totalEventsToGenerate = 3 * messageCount;
+    TestStreamTask.endLatch = new CountDownLatch(totalEventsToGenerate);
+
+    // create first processor
+    CountDownLatch startCountDownLatch1 = new CountDownLatch(1);
+    StreamProcessor sp = createStreamProcessor("1", map, startCountDownLatch1, null);
+
+    // produce first batch of messages starting with 0
+    produceMessages(0, inputTopic, messageCount);
+
+    // start the first processor
+    Thread t1 = runInThread(sp, TestStreamTask.endLatch);
+    t1.start();
+
+    // wait until the processor reports that it has started
+    try {
+      startCountDownLatch1.await(1000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Assert.fail("got interrupted while waiting for the first processor to start.");
+    }
+
+    // make sure it consumes all the messages from the first batch
+    waitUntilConsumedN(totalEventsToGenerate - messageCount);
+
+    // start the second processor
+    CountDownLatch countDownLatch2 = new CountDownLatch(1);
+    StreamProcessor sp2 = createStreamProcessor("2", map, countDownLatch2, null);
+    Thread t2 = runInThread(sp2, TestStreamTask.endLatch);
+    t2.start();
+
+    // wait until the processor reports that it has started
+    try {
+      countDownLatch2.await(1000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Assert.fail("got interrupted while waiting for the 2nd processor to start.");
+    }
+
+    // wait for at least one full debounce time to let the system to publish and distribute the new job model
+    TestZkUtils.sleepMs(3000);
+
+    // produce the second batch of the messages, starting with 'messageCount'
+    produceMessages(messageCount, inputTopic, messageCount);
+
+    // wait until all the events are consumed
+    // make sure it consumes all the messages from the first batch
+    waitUntilConsumedN(0);
+
+    // shutdown both
+    try {
+      synchronized (t1) {
+        t1.notify();
```

Comment:
this is the way to stop the thread. See runThread in the TestZkStreamProcessorBase.java